[clang-tidy] Return error code on config parse error

galenelias opened the pull request:
This addresses:

Issue #77341: clang-tidy fails to parse config files silently, even with --verify-config.

Currently when clang-tidy attempts to parse a .clang-tidy config file with invalid syntax, it emits and error and moves on, which can often result in applying the default set of checks, which is undesirable.

This change bubbles up the error state from parsing the configuration files up to `clangTidyMain` so we can do an early return with an error code. It's unfortunately quite a bit of plumbing - but this seemed to be the idiomatic approach for the codebase.

galenelias committed:
[clang-tidy] Return error code on config parse error

This addresses:

Issue 77341: clang-tidy fails to parse config files silently, even with
--verify-config.

Currently when clang-tidy attempts to parse a .clang-tidy config file
with invalid syntax, it emits and error and moves on, which can often
result in applying the default set of checks, which is undesirable.

This change bubbles up the error state from parsing the configuration
files up to `clangTidyMain` so we can do an early return with an error
code. It's unfortunately quite a bit of plumbing - but this seemed to be
the idiomatic approach for the clang-tidy codebase.

carlosgalvezp commented:
This is great, thank you! I'll have a look in the coming days.

carlosgalvezp commented on clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp:
```diff
@@ -266,7 +266,7 @@ ClangTidyOptions ClangTidyContext::getOptionsForFile(StringRef File) const {
   // unset values.
   return ClangTidyOptions::getDefaults().merge(
-      OptionsProvider->getOptions(File), 0);
+      *OptionsProvider->getOptions(File), 0);
```
How do we know if we can safely dereference without checking if we have an error?

galenelias replied:
My analysis was that because clangTidyMain does the validation explicitly up front and returns an error if it can't parse the config, that once it runs through here, we are basically guaranteed that it will succeed, because the errors have been caught.

However, digging in a bit deeper, I think there are at least two holes in this understanding: clangTidyMain only pre-validates the config for the first file passed in, so passing in a second file in a different directory with an invalid configuration will crash here.  The second is potentially the 'ClangTidyPlugin' code, which seems to be another entry point into the clang tidy code that bypasses clangTidyMain.  I'm not sure what the ClangTidyPlugin is or how to validate its behavior in the face of an invalid config.

I'll need to do a bit more digging to figure out the right strategy here.  Just bubbling up ErrorOr returns doesn't always seem great, but might unfortunately be required.

galenelias replied:
Ok, so my current plan is:
* Loop through all input files in clangTidyMain to explicitly validate configs for each input file and be able to return an error code there if a failure is found.
* In this method (`ClangTidyContext::getOptionsForFile`) just return the default config options if an error is found, which is similar to the behavior before this PR.  It seems like trying to make getOptionsForFile return an ErrorOr object and handle that at each call site is very cumbersome and leads to a lot of non-obvious failure modes that need to be resolved.

carlosgalvezp commented on clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:
```diff
@@ -337,8 +337,7 @@ Allow empty enabled checks. This suppresses
 )"),
-                                         cl::init(false),
-                                         cl::cat(ClangTidyCategory));
+                                   cl::init(false), cl::cat(ClangTidyCategory));
```
This formatting change seems unrelated, please revert

galenelias replied:
Oops, sorry.  It would be nice if the files were 'clang-format clean' as I am used to a flow where I can just run `clang-format -i` on all my modified files to ensure I'm following the formatting guidelines.  That is unfortunately not the case here.

carlosgalvezp commented on clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:
```diff
@@ -370,8 +369,8 @@ static void printStats(const ClangTidyStats &Stats) {
-static std::unique_ptr<ClangTidyOptionsProvider> createOptionsProvider(
-   llvm::IntrusiveRefCntPtr<vfs::FileSystem> FS) {
+static std::unique_ptr<ClangTidyOptionsProvider>
+createOptionsProvider(llvm::IntrusiveRefCntPtr<vfs::FileSystem> FS) {
```
This formatting change seems unrelated, please revert

carlosgalvezp commented on clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:
```diff
@@ -624,21 +623,29 @@ int clangTidyMain(int argc, const char **argv) {
-                       << ".\n";
-          break;
+
+    if (RawOptions) {
```
if (!RawOptions) return 1?

carlosgalvezp commented on clang-tools-extra/unittests/clang-tidy/OptionsProviderTest.cpp:
```diff
@@ -46,20 +46,20 @@ TEST(ClangTidyOptionsProvider, InMemoryFileSystems) {
   // 2 and 3 should use the same config so these should also be the same.
-  EXPECT_EQ(File2Options.Checks, File3Options.Checks);
+  EXPECT_EQ(File2Options->Checks, File3Options->Checks);
 }
```
Missing a unit test that demonstrates that clang-tidy exists early with exit code 1 when the config file could not be parsed/found

galenelias replied:
I couldn't find any tests which exercise `clangTidyMain` (which makes sense, as that is not really the layer a unit test would operate on.  The best place I could find to put a test was near the ClangTidyOptionsProvider.InMemoryFileSystems test, which at least tests the behavior with multiple configuration files, as well as some amount of the propagation logic.

I will go ahead and add a test there unless there is a more appropriate spot for verifying the high level "early exit code 1" behavior of the PR.

carlosgalvezp reviewed (changes requested):
Please document the change in the Release Notes

galenelias committed:
Add Release Notes

galenelias committed:
Add some Unit Test coverage

galenelias committed:
Revert unrelated clang-format fixes

galenelias committed:
Return error code in ExplainConfig path

galenelias committed:
Validate configs for all inputs files up front

And make ClangTidyContext::getOptionsForFile resilient to errors at that
point.

galenelias commented:
Ping

carlosgalvezp commented:
Apologies for the delay, I will review during this week!

carlosgalvezp commented on clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp:
```diff
@@ -265,8 +265,16 @@ const ClangTidyOptions &ClangTidyContext::getOptions() const {
+  llvm::ErrorOr<ClangTidyOptions> fileOptions =
+      OptionsProvider->getOptions(File);
+
+  // If there was an error parsing the options, just use the default options.
```
Shouldn't we fail hard/propagate the error further here? Returning the default options hides the problem for the user.

galenelias replied:
I agree, but it seemed quite difficult to propagate an error from this function and would drastically increase the size and invasiveness of this change, as every call of this function would need to propagate the error, and so on and so forth recursively until dozens+ of functions have been changed.

Given that in the primary clang-tidy usage scenario that these configs have already been pre-validated, and that this fallback behavior is basically equivalent to the previous behavior of silently continuing on with the default config, this seemed like a reasonable compromise.

I can pursue propagating the error, but when I initially investigated it, it seems quite complicated given the current places this method is called and how prepared the callers were for propagating errors.

carlosgalvezp commented on clang-tools-extra/unittests/clang-tidy/OptionsProviderTest.cpp:
```diff
@@ -46,20 +46,65 @@ TEST(ClangTidyOptionsProvider, InMemoryFileSystems) {
-  EXPECT_EQ(*File1Options.Checks, "-*,clang-diagnostic-*,readability-*");
-  ASSERT_TRUE(File2Options.Checks.has_value());
-  EXPECT_EQ(*File2Options.Checks, "bugprone-*,misc-*,clang-diagnostic-*");
+  ASSERT_TRUE(File1Options->Checks.has_value());
```
Before dereferencing, we should assert that there was no error

galenelias commented on clang-tools-extra/unittests/clang-tidy/OptionsProviderTest.cpp:
```diff
@@ -46,20 +46,65 @@ TEST(ClangTidyOptionsProvider, InMemoryFileSystems) {
+  llvm::ErrorOr<ClangTidyOptions> File3Options =
+      FileOpt.getOptions("ProjectRoot/SubDir1/SubDir2/File.cpp");
+
+  ASSERT_TRUE(!File1Options);
```
Yah, sorry - wasn't being very idiomatic here.  Cleaned them up.  I don't think we need the negations or double negations to coerce to booleans, it seems like we can just use the ASSERT_ macros directly.  Not sure why I formulated them this way.  Apologies.

galenelias committed:
Address review comments

galenelias committed:
Fix formatting of test header comment

galenelias committed:
Merge remote-tracking branch 'upstream/main' into user/gelias/issue77341_clangTidyBubbleConfigParseError

galenelias committed:
Merge remote-tracking branch 'upstream/main' into user/gelias/issue77341_clangTidyBubbleConfigParseError

galenelias committed:
Clang-format and fix merge changes

tiagomacarios commented:
@carlosgalvezp Any updates?
